internal/abi/fastly/hostcalls_guest.go

```diff
@@ -1084,15 +1095,21 @@ func fastlyHTTPReqURIGet(
 ) FastlyStatus
 
 // GetURI returns the fully qualified URI of the request.
-func (r *HTTPRequest) GetURI(maxURLLen int) (string, error) {
-	buf := prim.NewWriteBuffer(maxURLLen)
-
-	if err := fastlyHTTPReqURIGet(
+func (r *HTTPRequest) GetURI() (string, error) {
+	n := DefaultMediumBufLen // Longest (unknown); Typically less than 1024, but some browsers accept much longer
+alloc:
+	buf := prim.NewWriteBuffer(n)
+	status := fastlyHTTPReqURIGet(
 		r.h,
 		prim.ToPointer(buf.Char8Pointer()),
 		buf.Cap(),
 		prim.ToPointer(buf.NPointer()),
-	).toError(); err != nil {
+	)
+	if status == FastlyStatusBufLen && buf.NValue() > 0 {
+		n = int(buf.NValue())
+		goto alloc // goto saves having to allocate a function closure and avoids having to duplicate the hostcall
+	}
+	if err := status.toError(); err != nil {
```

Comment:
All these methods are copied all over, so maybe we could refactor them by providing a function that takes a `func` that actually makes the hostcall, and handles the adaptive buffer handling?

Then these API entrypoints would look something like this:

```go
func (r *HTTPRequest) GetSomething() (string, error) {
     n := DefaultMediumBufLen // explanation why this is the default here
     if buf, err := hostcallWithAdaptiveBufferSize(n, fastlyHTTPReqSomethingGet); err != nil {
           return nil, err
     }
     return buf.ToString(), nil
}
```

Reply:
I'd love to do this but the type system will fight us on passing the various hostcalls, since they all have different signatures. To get around that we need a closure like `func() { fastlyHostcall(args, args, args) }` and then I think we start losing the allocation battle.

Reply:
Yeah, you're right.  Tinygo isn't great at figuring out closures are essentially forwarding arguments to another function and generating something better. :(